### PR TITLE
textdecoder

### LINKED ISSUE #4402: Use TextDecoder to optimize decoding C strings to JS
There is a recent API for this,

https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/decode

Looks like it's currently in firefox and chrome.

We could check if it's present and use it if so, instead of the probably much slower code we have now that does it manually in JS, which is in `Pointer_stringify` in `src/preamble.js` (there is TODO there for this).

### PR DESCRIPTION
**Pull request by juj**
Use TextDecoder to decode UTF-8 and UTF-16 strings. Closes #4402.

Not quite yet to merge, needs the fallback path testing and a bit more benchmarking.

### CODE REVIEW

File: src/preamble.js [hunk 1]
```diff
@@ -515,39 +515,49 @@ function stringToAscii(str, outPtr) {
 // Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
 // a copy of that string as a Javascript String object.
 
+var utf8_decoder = new TextDecoder('utf8');
```
Comment (kripken) [inline src/preamble.js]: maybe `UTF8Decoder` for consistency?

File: src/preamble.js [hunk 2]
```diff
@@ -515,39 +515,49 @@ function stringToAscii(str, outPtr) {
...
 function UTF8ArrayToString(u8Array, idx) {
-  var u0, u1, u2, u3, u4, u5;
+  var endPtr = idx;
+  // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
+  // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
+  while(u8Array[endPtr]) ++endPtr;
```
Comment (kripken) [inline src/preamble.js]: space after `while`

File: src/preamble.js [hunk 3]
```diff
@@ -669,17 +679,27 @@ function lengthBytesUTF8(str) {
 // Given a pointer 'ptr' to a null-terminated UTF16LE-encoded string in the emscripten HEAP, returns
 // a copy of that string as a Javascript String object.
 
+var utf16_decoder = new TextDecoder('utf-16le');
 function UTF16ToString(ptr) {
-  var i = 0;
+  var endPtr = ptr;
+  // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
+  // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
+  while({{{ makeGetValue('endPtr', 0, 'i16') }}}) endPtr += 2;
```
Comment (kripken) [inline src/preamble.js]: space after `while`

File: src/preamble.js [hunk 4]
```diff
@@ -669,17 +679,27 @@ function lengthBytesUTF8(str) {
...
-    str += String.fromCharCode(codeUnit);
+  if (endPtr - ptr > 32) {
+    return utf16_decoder.decode(HEAPU8.subarray(ptr, endPtr));
+  } else {
+    var i = 0;
+
+    var str = '';
+    while (1) {
+      var codeUnit = {{{ makeGetValue('ptr', 'i*2', 'i16') }}};
+      if (codeUnit == 0)
```
Comment (kripken) [inline src/preamble.js]: same line or `{}` (looks like this was an error already present)

**Comment by juj** [comment]
Addressed review comments and added tests. Also added an option `-s TEXTDECODER=0/1` to allow disabling the feature if needed. This is mainly to allow easy performance testing for users in their environments, since it is difficult to offer conclusive testing in a synthetic test environment for UTF strings (e.g. some workloads might have mostly 7-bit ASCII strings, whereas others could be 4-byte UTF-8 heavy)

**Comment by kripken** [comment]
Overall lgtm. The one concern I have are the new alignment requirements this adds. This might break existing code, although I guess very rarely?

**Comment by juj** [comment]
Actually, the alignment requirements are old and have been present already earlier, see e.g. https://github.com/kripken/emscripten/blob/master/src/preamble.js#L677 that does a HEAP16 load. Only when updating this PR I realized we have this requirement, and added the asserts in for diligence. It would be possible to add support for unaligned cases, but probably not critical for now.

**Comment by kripken** [comment]
I see, thanks. Yeah, since this was the case before, it all sounds good.

**Pull request merged**